pkg/storage/scrub.go

```diff
@@ -0,0 +1,395 @@
+
+	colImageNameIndex = 0
+	colTagIndex       = 1
+	colStatusIndex    = 2
+	colErrorIndex     = 3
+
+	sec = 5
+)
+
+type checksPool struct {
+	jobs      chan *checkJob
+	done      chan struct{}
+	waitGroup *sync.WaitGroup
+	outputCh  chan stringResult
+	context   context.Context
+}
+
+type checkJob struct {
+	dir       string
+	imageName string
+	tag       string
+	ctx       context.Context
+	oci       casext.Engine
+	manifest  ispec.Descriptor
+}
+
+type stringResult struct {
+	StrValue string
+	Err      error
+}
+
+const rateLimiterBuffer = 100
+
+func newSmoothRateLimiter(ctx context.Context, wg *sync.WaitGroup, op chan stringResult) *checksPool {
+	ch := make(chan *checkJob, rateLimiterBuffer)
+
+	return &checksPool{
+		jobs:      ch,
+		done:      make(chan struct{}),
+		waitGroup: wg,
+		outputCh:  op,
+		context:   ctx,
+	}
+}
+
+// block every "rateLimit" time duration.
+const rateLimit = 100 * time.Millisecond
+
+func (p *checksPool) startRateLimiter() {
+	p.waitGroup.Done()
+
+	throttle := time.NewTicker(rateLimit).C
+
+	for {
+		select {
+		case job := <-p.jobs:
+			go p.doJob(job)
+		case <-p.done:
+			return
+		}
+		<-throttle
+	}
+}
+
+func (p *checksPool) doJob(job *checkJob) {
+	defer p.waitGroup.Done()
+
+	// check manifest and config
+	stat, err := umoci.Stat(job.ctx, job.oci, job.manifest)
+
+	var str string
+
+	if err != nil {
+		str = job.printResult(err)
+	} else {
+		// check layers
+		for _, s := range stat.History {
+			layer := s.Layer
+			if layer == nil {
+				continue
+			}
+
+			// check if layer exists
+			bytes, err := ioutil.ReadFile(path.Join(job.dir, "blobs/sha256", layer.Digest.Hex()))
+			if err != nil {
+				str = job.printResult(errors.ErrBlobNotFound)
+				break
+			}
+
+			// check layer content
+			expectedHash := fmt.Sprintf("%x", sha256.Sum256(bytes))
+			if expectedHash != layer.Digest.Hex() {
+				str = job.printResult(errors.ErrBadBlobDigest)
+				break
+			}
+			str = job.printResult(nil)
+		}
+	}
+
+	if isContextDone(p.context) {
+		return
+	}
+
+	p.outputCh <- stringResult{str, nil}
+}
+
+func (p *checksPool) submitJob(job *checkJob) {
+	p.jobs <- job
+}
+
+func isContextDone(ctx context.Context) bool {
+	select {
+	case <-ctx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+func addManifestCallToPool(ctxUmoci context.Context, p *checksPool, imageName, tagName string,
+	wg *sync.WaitGroup, oci casext.Engine, m ispec.Descriptor, dir string) {
+	defer wg.Done()
+
+	job := checkJob{
+		dir:       dir,
+		imageName: imageName,
+		tag:       tagName,
+		ctx:       ctxUmoci,
+		oci:       oci,
+		manifest:  m,
+	}
+
+	wg.Add(1)
+	p.submitJob(&job)
+}
+
+func (sc StoreController) CheckAllBlobsIntegrity(resultWriter io.Writer) error {
+	imageErr := make(chan stringResult)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	var wg sync.WaitGroup
+
+	imagesList := []ImageStore{}
+	imagesList = append(imagesList, sc.DefaultStore)
+
+	for _, subStore := range sc.SubStore {
```

Comment:
This design may just be overkill. 
Why not just iterate over the repos in a simple tight loop as first implementation. I don't think you need anything more than that. When you make it online you can move that to a single goroutine and pls pay attention to the locks. Let's not optimize prematurely.

Reply:
Ok. I changed the implementation and now it just iterates over the repos in a loop.